### ISSUE TITLE
Support for IOS 13 notifications

### ISSUE THREAD

**Issue by cesarve77** [node-apn/node-apn#666]
Now apn need to include a new param apns-push-type

https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns?language=objc

**Comment by ninefivetwo** [comment]
This is has already been implemented in #656.
Check the docs for [pushType](https://github.com/node-apn/node-apn/blob/master/doc/notification.markdown#notificationpushtype).

But the issue is that it is currently waiting to be published..

**Comment by vivek8690** [comment]
VOIP push notifications are not working after updating to iOS 13

**Comment by alemar11** [comment]
Update: using this (fork)[https://github.com/parse-community/node-apn] with apns-push-type to "voip" is working again!

**Comment by vivek8690** [comment]
By using [https://github.com/parse-community/node-apn] not working for me @alemar11 

TypeError: apn.Device is not a constructor

**Comment by vivek8690** [comment]
@ninefivetwo pushType changes are updated on npm package because the published version is too old

**Comment by jackfurr** [comment]
Any chance that this can be published to npm?  When I `npm install apn` it doesn't include latest master with 'pushType' support.

**Comment by vivek8690** [comment]
Notification.prototype.headers = function headers() {
  let headers = {};

  if (this.priority !== 10) {
    headers["apns-priority"] = this.priority;
  }

  if (this.id) {
    headers["apns-id"] = this.id;
  }

  if (this.expiry >= 0) {
    headers["apns-expiration"] = this.expiry;
  }

  if (this.topic) {
    headers["apns-topic"] = this.topic;
  }
  if (this.pushType) {
    headers["apns-push-type"] = this.pushType
  }

  if (this.collapseId) {
    headers["apns-collapse-id"] = this.collapseId;
  }
  return headers;
};

**Comment by vivek8690** [comment]
@jackfurr you can replace with above-mentioned code in lib/notifaction/index.js. for now

**Comment by jackfurr** [comment]
> @jackfurr you can replace with above-mentioned code in lib/notifaction/index.js. for now

Thanks, but this won't work for my build process that pulls from npm (`npm install`).  I can temporarily fork this and deploy to private registry until this is pushed to npm.... or use a different module.

Is there an issue with pushing to npm?  Just trying to understand.

**Comment by RichAppz** [comment]
There is an issue, the `npm install apn` installs 2.2.0 not 3.0.0

**Comment by markschmid** [comment]
https://github.com/node-apn/node-apn/issues/664

**Comment by tangkhaiphuong** [comment]
Please release it.

**Comment by naveenreddy3** [comment]
Any update on this?

**Comment by iamromec** [comment]
Even using 3.0.0 as dependencies does not resolve the issue. I've also tried using "@parse/node-apn": "^3.1.0", but the issue is still there. 

Guys, we need support in this. It's really a pathetic change by Apple and punished a huge number of apps. I also cannot update the version of phonegap-push-plugin (for ionic) because it doubles the size.

Help here out.

**Comment by dplewis** [comment]
Parse has a published module

https://github.com/parse-community/node-apn

You should be able to install from the master branch of this repo.

**Comment by iamromec** [comment]
@dplewis I have tried that but still getting 'BadDeviceToken' issue while sending the notification with Parse community module.

**Comment by dplewis** [comment]
Do you use certs or a single key?

**Comment by iamromec** [comment]
I use the key. Also also sent pushType='alert'.

apnNotification = new apn.Notification();
apnNotification.sound = 'ping.aiff';
 apnNotification.body = data.body;
 apnNotification.pushType = 'alert';

**Comment by iamromec** [comment]
Do you have an update? It's really becoming worse.

**Comment by chucktee** [comment]
@iamromec Check that the device token you are using is being properly formatted. If you are using the deviceToken.description, as many were, you will now get BadDeviceToken as iOS 13 reports this differently.

**Comment by iamromec** [comment]
Yeah I'm receveing the new device token that is {length=32,bytes=0x50b570943687112b212d4baf36897589...7asdwwdwqd121323d}

I tried decoding it with: Buffer.from(b64string, 'base64'); but didn't work.

Could you please guide how I can get it decoded with Node.js because I don't want to upgrade the phonegap-push-plugin 3.0.0

**Comment by iamromec** [comment]
Guys? I need help here....pissed off looking for solution for last 3 days.

**Comment by chucktee** [comment]
Google is your friend. Maybe "convert device token to a string" or "decode iOS 13 device token"

**Comment by dplewis** [comment]
@iamromec As @chucktee mention the reason could be why you are getting the new device token or decode the new device token.

Here is a [SO Thread](https://stackoverflow.com/questions/57839723/does-ios-13-has-new-way-of-getting-device-notification-token) documenting that

**Comment by RichAppz** [comment]
I had to move away from this repo and use `@parse/node-apn` as this does not pull version 3.0.0 and wasn't receiving the support required to get this up and running.

**Comment by RichAppz** [comment]
Also if you are having issues with watchOS APNs tokens then you can check out the article I wrote after spending 2 weeks trying to get to the bottom of the issues I was having: https://medium.com/@RichMucha/independent-watchos-and-apns-4a8ede6f44e8

**Comment by iamromec** [comment]
@chucktee, @dplewis  I found solutions for Swift or C# but cannot find a solution for Node.js. Could you please suggest how do we decode it in Node.js?

**Comment by dplewis** [comment]
@iamromec why not open an issue in the plugin your using? The solution is probably there or you can debug it yourself. You should also post the swift and C# links you found. That community can help you.

**Comment by iamromec** [comment]
I've opened the issue over there too: https://github.com/phonegap/phonegap-plugin-push/issues/2853

The problem with the latest version of the plugin is it increases the app size and slows down the performance like hell which I cannot accept at all. That's why we need a solution at the backend level to decode it or get node-apn do it.

**Comment by chucktee** [comment]
@iamromec (I'm guessing your iOS side is written in Phonegap?) If you've already found a Swift / Objective-C solution, just do it on the device when the device registers for notifications. Phonegap has support for calling native code.

**Comment by iamromec** [comment]
Wooooooh! Finally, I've fixed it. I'll write an article so that it might help others. I actually played around the Swift code within the plugin and replaced some blocks with the latest version of the phonegap-push-plugin.

**Comment by iamromec** [comment]
I think I've fixed it but waiting for the app is waiting for review so. I'll update you on what I did to fix it without increasing the app size and slowing down the performance.

**Comment by wiggisser** [comment]
@iamromec There is no way, you can get the correct device token from a string like `{length=32,bytes=0x50b570943687112b212d4baf36897589...7asdwwdwqd121323d}` because there are essential parts of the token (ie where the ... are located) missing. This has to be done on the app's side

**Comment by iamromec** [comment]
@wiggisser you're right. People kept talking about on the internet that it's an encoded token and that created a false impression. It cannot be decoded. The problem is not with the library nor there is a need of updating to 3.0.0.

The solution lies within the Swift code of retrieving device token:

For Cordova applications:

Go to: plugins/phonegap-plugin-push/src/ios/PushPlugin.m and replace (can be found in the _didRegisterForRemoteNotificationsWithDeviceToken()_ function):

_**NSString *token = [[[[deviceToken description] stringByReplacingOccurrencesOfString:@"<"withString:@""]
                        stringByReplacingOccurrencesOfString:@">" withString:@""]
                       stringByReplacingOccurrencesOfString: @" " withString: @""];**_

with

_#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 130000
    // [deviceToken description] is like "{length = 32, bytes = 0xd3d997af 967d1f43 b405374a 13394d2f ... 28f10282 14af515f }"
    NSString *token = [self hexadecimalStringFromData:deviceToken];
#else
    // [deviceToken description] is like "<124686a5 556a72ca d808f572 00c323b9 3eff9285 92445590 3225757d b83967be>"
    NSString *token = [[[[deviceToken description] stringByReplacingOccurrencesOfString:@"<"withString:@""]
                        stringByReplacingOccurrencesOfString:@">" withString:@""]
                       stringByReplacingOccurrencesOfString: @" " withString: @""];
#endif_

and add this function to this same file:

_**(NSString` *)hexadecimalStringFromData:(NSData *)data
{
    NSUInteger dataLength = data.length;
    if (dataLength == 0) {
        return nil;
    }
    const unsigned char *dataBuffer = data.bytes;
    NSMutableString *hexString  = [NSMutableString stringWithCapacity:(dataLength * 2)];
    for (int i = 0; i < dataLength; ++i) {
        [hexString appendFormat:@"%02x", dataBuffer[i]];
    }
    return [hexString copy];
}**_


Then remove the platform ios and add again and then rebuild.

It should fix the problem.

**Comment by yogeshbansal** [comment]
@iamromec : Hi could you please tell me what else you have done to fix the issue. After making changes in both cordova plugin and nodejs side. I am able to get the token in decoded format, but when i am trying to send push notification , its failed with **BadDeviceToken** error.

**Comment by wiggisser** [comment]
@yogeshbansal is your app production or development? Do you send the request to the correct endpoint?

**Comment by iamromec** [comment]
@yogeshbansal did you follow the steps I've mentioned in the https://github.com/node-apn/node-apn/issues/666#issuecomment-551842757 comment?

**Comment by iamromec** [comment]
@yogeshbansal It's not an encoded string. It's an encoded string with some part of it is missing so you can not decode it. You need to make changes in the front-end, not backend.

**Comment by yogeshbansal** [comment]
> @yogeshbansal is your app production or development? Do you send the request to the correct endpoint?

Its production app, yes it was working before

**Comment by yogeshbansal** [comment]
Yes, I made the changes is front end, in the backend i have added the property pushType= 'alert'. I think i am getting the token in correct format. But still its failing..any other thoughts

**Comment by wiggisser** [comment]
@yogeshbansal You "think" you are receiving the token in the correct format? So how does the token look like on the backend? It must be a hexadecimal string of length 64 without any spaces or any other characters. Ie it must match the following regex /^[0-9a-f]{64}$/

**Comment by yogeshbansal** [comment]
this is how token look like in the backend 'a4c6df0254d1c5f67c2b3426397cbe64895f442ef6025d0cb07f31a4459707e5'

**Comment by wiggisser** [comment]
@yogeshbansal This has at least the correct format for a device token. If you still get the BadDeviceToken error you should verify, you are sending your request to the correct apns endpoint. Ie check if your options when creating the  `var provider = new apn.Provider(options)` contain the correct value for the `production` option. As your iOS app is production, be aware, that the default value for the `production` option is `false`, ie requests are sent to the sandbox environment, unless there is an environment variable `NODE_ENV=production` is set (https://github.com/node-apn/node-apn/#connecting). 

If the production option is set correctly, only apple support may be able to help you finding the reason why the token is rejected. That's not an issue of this library then.

**Comment by iamromec** [comment]
"@wiggisser this is how token look like in the backend 'a4c6df0254d1c5f67c2b3426397cbe64895f442ef6025d0cb07f31a4459707e5'"

--> It's not the correct token. It's just a response after 'new Buffer()' and its string version. 

The fact is the token you are getting {length=32,bytes=0x67dafb604ef75b082feedf2fa2aa01f2...b7bb7bb7bsdfsdfsdfsdf} is not the encryption of any string or token. It is but some part is missing '...' so you cannot decrypt it at all. 

For this either upgrade your cordova plugin or follow the method I've mentioned above. 
-> https://github.com/node-apn/node-apn/issues/666#issuecomment-551842757

**Comment by wiggisser** [comment]
@iamromec 
Why are you so sure, `a4c6df0254d1c5f67c2b3426397cbe64895f442ef6025d0cb07f31a4459707e5` is not a correct token? It has the correct format (ie a 64 character hex string) and @yogeshbansal said he already did make changes in the frontend. 

In fact what you are getting from iOS in the `didRegisterForRemoteNotificationsWithDeviceToken` handler is a databuffer with 32 byte of data. And you have to encode those 32 byte of data as a 64 char hex string (two characters for one byte), which will be recognized as device token. The above string is a 64 char hex string, so nobody except apple can tell, whether this is a valid device token or not.

**Comment by imromec** [comment]
@wiggisser Just use this token and send a notification if you receives it that means it's a correct token. If you don't receive that means it's not a valid token.

**Comment by wiggisser** [comment]
@imromec 

> @wiggisser Just use this token and send a notification if you receives it that means it's a correct token. If you don't receive that means it's not a valid token.

Unfortunately it's not that easy. If you for instance have an app build in dev-configuration and you send the token created by this app to the apns production endpoint (or vice versa), you will also receive a "BadDeviceToken" error, even if the token itself is perfectly valid, just not for that environment.

**Comment by BorntraegerMarc** [comment]
Just my 2 cents: We've been using VoIP push for iOS 13 devices just fine. However you have to distribute the app with XCode 10.X to the app store. If you build the app with XCode 11 then iOS will crash if it receives a VoIP push which doesn't show the call UI.

**Comment by iamromec** [comment]
It's again! Now it's showing {403: InvalidProviderToken} where the token is correct 100% sure. WHAT'S WRONG WITH iOS notifications? It's really frustrating.

**Comment by yoneapp** [comment]
iOS 13.3 has a bug.
Push notifications may be skipped.
Apple says it fixed it in 13.3.1.
For your reference:)

**Comment by BorntraegerMarc** [comment]
> iOS 13.3 has a bug.
Push notifications may be skipped. (wifi)
Apple says it fixed it in 13.3.1.

Interesting. Do you have an official source on that? (For reference)

**Comment by yoneapp** [comment]
> Resolves an issue where push notifications could fail to be delivered over Wi-Fi
> https://support.apple.com/en-us/HT210393#1331

Please refer to the link.

**Comment by tapz** [comment]
Is 3.0.0 still alpha? The only 3 version found from NPM is "3.0.0-alpha1".
